clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -402,26 +403,38 @@ public void refreshCommittedOffsetsIfNeeded() {
         }
     }
 
-    @Override
-    public void close() {
+    public void close(long timeoutMs, long requestTimeoutMs) {
         // we do not need to re-enable wakeups since we are closing already
         client.disableWakeups();
+        long startTimeMs = time.milliseconds();
         try {
-            maybeAutoCommitOffsetsSync();
+            timeoutMs = Math.min(timeoutMs, requestTimeoutMs);
+            maybeAutoCommitOffsetsSync(timeoutMs);
+        } finally {
+            super.close();
+        }
 
-            Node coordinator;
-            long endTimeMs = time.milliseconds() + CLOSE_TIMEOUT_MS;
-            while ((coordinator = coordinator()) != null && client.pendingRequestCount(coordinator) > 0) {
-                long remainingTimeMs = endTimeMs - time.milliseconds();
-                if (remainingTimeMs > 0)
+        Node coordinator = coordinator();
+        long now = time.milliseconds();
+        long endTimeMs = now + Math.min(timeoutMs - (now - startTimeMs), requestTimeoutMs);
+        while (pendingAsyncCommits.get() > 0 || (coordinator != null && client.pendingRequestCount(coordinator) > 0)) {
```

Comment:
The `pendingRequestCount` check ensures that `LeaveGroup` is actually sent. We could use `unsent` count from `ConsumerNetworkClient` instead, but since `LeaveGroup` is not sent if the coordinator is not known and `pendingRequestCount` is only checked when coordinator is known, waiting for `LeaveGroup` response shouldn't have too much impact on the overall close time.